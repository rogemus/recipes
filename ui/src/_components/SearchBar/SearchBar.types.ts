import type { Comp } from "@/_models";

export type SearchBarProps = {} & Comp;
