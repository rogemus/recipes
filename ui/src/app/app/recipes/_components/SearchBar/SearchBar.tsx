"use client";

import { useState, type ChangeEvent, type FC } from "react";
import { useDebouncedCallback } from "use-debounce";

import { getAutocompleteRecipes } from "../../_lib/actions";
import { SearchResults } from "../SearchResults";

import type { RecipeSimple } from "@/_models";
import type { SearchBarProps } from "./SearchBar.types";

const SearchBar: FC<SearchBarProps> = ({ testId = "SearchBar" }) => {
  const [results, setResults] = useState<RecipeSimple[]>([]);
  const fetchRecipes = async (query: string) => {
    try {
      const response = await getAutocompleteRecipes(query);

      if (response.data) {
        setResults(response.data.recipes);
      }
    } catch (e) {
      setResults([]);
      console.log("ERROR: cannot fetch recipes");
    }
  };

  const handleInputChange = useDebouncedCallback(
    (e: ChangeEvent<HTMLInputElement>) => {
      if (e.target.value.length >= 3) {
        fetchRecipes(e.target.value);
      }
    },
    300,
  );

  return (
    <div data-testid={testId}>
      <div>
        <label>Search recipe</label>
        <input
          type="text"
          data-testid={`${testId}-input`}
          onChange={handleInputChange}
          placeholder="Search recipe..."
        />
      </div>
      <SearchResults recipes={results} />
    </div>
  );
};

export { SearchBar };
