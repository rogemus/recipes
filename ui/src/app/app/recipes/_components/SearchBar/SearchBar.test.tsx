import { expect, test, describe } from "vitest";
import { render } from "@testing-library/react";

import { SearchBar } from "./SearchBar";

describe("@component/SearchBar", () => {
  test("renders component correctly", () => {
    const { container } = render(<SearchBar />);

    expect(container).toMatchSnapshot();
  });
});
