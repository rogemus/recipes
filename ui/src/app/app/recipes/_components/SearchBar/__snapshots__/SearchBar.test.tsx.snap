// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@component/SearchBar > renders component correctly 1`] = `
<div>
  <div
    data-testid="SearchBar"
  >
    <div>
      <label>
        Search recipe
      </label>
      <input
        data-testid="SearchBar-input"
        placeholder="Search recipe..."
        type="text"
      />
    </div>
  </div>
</div>
`;
